### ISSUE TITLE
Reuse a dynamic buffer for different draw calls

### ISSUE THREAD

**Issue by bvssvni** [PistonDevelopers/glium_graphics#5]
Currently it creates a new vertex buffer at each call.

It also allocates in `.collect()` when converting from a flat array to the buffer.

**Comment by tomaka** [comment]
For performance, it should also use a persistent buffer when available.

[!http://ouiouiphoto.fr/Wp/wp-content/uploads/2014/08/Flash.jpg]()

**Comment by bvssvni** [comment]
gfx_graphics uses `graphics::BACK_END_MAX_VERTEX_COUNT` to compute the size of the buffer upfront, which piston-graphics uses to limit the chunk size. This might be changed in the future to reduce cache misses.

**Comment by aepsil0n** [comment]
Removing the allocation would require Glium's buffers to be constructed from slices, I think.

**Comment by bvssvni** [comment]
Is it possible to update the buffer with slices?

**Comment by tomaka** [comment]
It's not a problem of slice vs vec. Creating a vertex buffer is a slightly costly operation due both to the implementation of glium (the background thread thing) and to https://github.com/tomaka/glium/issues/478.

The most optimal implementation, in my opinion, uses two persistent buffers.
When you draw something, you map and write the data in the one of the buffers, then draw with it. Every time you draw, you alternate the buffer that you use.

This way you use the CPU to upload the data in the video memory while the GPU is drawing simultaneously.

**Comment by tomaka** [comment]
I can write the implementation, but we'd need something similar to gfx_graphics's `G2D`, ie. a struct that persists between draw calls where we can store the buffer.

That's API-related so I'm going to let it to you.

**Comment by bvssvni** [comment]
@tomaka Just go ahead if you know what to do. If it gets similar to gfx_graphics then that's fine.

**Comment by tomaka** [comment]
I'll need to fix https://github.com/tomaka/glium/issues/481 first

**Comment by tomaka** [comment]
I'm blocked on https://github.com/tomaka/glium/pull/482 which is blocked on https://github.com/rust-lang/rust/issues/22156

**Comment by bvssvni** [comment]
It seems this is already done? Closing.
